### PR TITLE
feat: Trie data structure using hashmap, predict_words method, class Trie

### PR DESCRIPTION
**Pull request by bharath000**
#### Description of Change
feat: Trie data structure using hashmap, predict_words method to recommend words based on a given prefix. 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: data_structures/trie_using_hashmap.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+ * words based on prefix.
+ * @details The Trie data structure is implemented using unordered map to use
+ * memory optimally, predict_words method is developed to recommend words based
+ * on a given prefix along with other methods insert, delete, search, startwith
+ * in trie.
+ * @see trie_modern.cpp for difference
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for iostream
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:15]:
```suggestion
#include <iostream>       /// for IO operations
```

File: data_structures/trie_using_hashmap.cpp [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+#include <memory>         /// for shared pointer
+#include <stack>          /// for stack
+#include <unordered_map>  /// for unordered map
+#include <vector>         /// for vector
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:19]:
```suggestion
#include <vector>         /// for std::vector
```

File: data_structures/trie_using_hashmap.cpp [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+ * memory optimally, predict_words method is developed to recommend words based
+ * on a given prefix along with other methods insert, delete, search, startwith
+ * in trie.
+ * @see trie_modern.cpp for difference
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for iostream
+#include <memory>         /// for shared pointer
+#include <stack>          /// for stack
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:17]: Please add proper documentation here.

File: data_structures/trie_using_hashmap.cpp [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+ * @details The Trie data structure is implemented using unordered map to use
+ * memory optimally, predict_words method is developed to recommend words based
+ * on a given prefix along with other methods insert, delete, search, startwith
+ * in trie.
+ * @see trie_modern.cpp for difference
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for iostream
+#include <memory>         /// for shared pointer
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:16]:
```suggestion
#include <memory>         /// for std::shared_ptr
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1464/checks?check_run_id=2098126392) warnings.

File: data_structures/trie_using_hashmap.cpp [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for IO operations
+#include <memory>         /// for std::shared_ptr
+#include <stack>          /// for std::stack
+#include <unordered_map>  /// for std::unordered_map
+#include <vector>         /// for std::vector
+
+/**
+ * @namespace data_structures
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:22]:
```suggestion
 * @namespace data_structures
 * @brief Data structures algorithms
```

File: data_structures/trie_using_hashmap.cpp [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+ */
+
+namespace data_structures {
+/**
+ * Class documentation
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:27]: Please add proper class documentation.

File: data_structures/trie_using_hashmap.cpp [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+            prefix);  ///< iteratively and recursively get the recommended words
+
+        return result;
+    }
+};
+
+}  // namespace data_structures
+
+/**
+ * @brief Test implementations
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:201]:
```suggestion
 * @brief Self-test implementations
```

File: data_structures/trie_using_hashmap.cpp [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+    std::cout << "No word starts with prefix h in trie" << std::endl;
+
+    std::cout << "All tests passed" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:325]:
```suggestion
```

File: data_structures/trie_using_hashmap.cpp (hunk 4 above)
Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: changes have been updated, thank you

File: data_structures/trie_using_hashmap.cpp [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+int main() {
+    test();
+
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:328]:
```suggestion
    test();  // run self-test implementations
```

File: data_structures/trie_using_hashmap.cpp [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+ */
+
+namespace data_structures {
+/**
+ * Class documentation
+ */
+class Trie {
+ private:
+    struct Node {
+        /** @brief trie nodes as key value pairs, value is a shared_pointer of
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:32]: Is this the documentation of the `Node` struct?

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/trie_using_hashmap.cpp (hunk 6 above)
Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: added documentation to class

File: data_structures/trie_using_hashmap.cpp (hunk 10 above)
Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: updated documentation to Node struct

File: data_structures/trie_using_hashmap.cpp [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+    assert(obj.search("about"));
+    assert(obj.search("abscond"));
+    assert(obj.search("bus"));
+    assert(obj.search("buses"));
+    assert(obj.search("Bounce"));
+    assert(obj.search("Apple"));
+
+    std::cout << "All the Inserted words are present in the trie" << std::endl;
+
+    /*test for startwith prefix method*/
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:251]:
```suggestion
    // test for startwith prefix method
```

File: data_structures/trie_using_hashmap.cpp [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+    assert(obj.startwith("bus"));
+    assert(obj.startwith("Bo"));
+    assert(obj.startwith("A"));
+    assert(!obj.startwith("Ca"));
+
+    assert(!obj.startwith("C"));
+
+    std::cout << "All the tests passed for startwith method" << std::endl;
+
+    /*test for predict_words/recommendation of words based on a given prefix*/
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:272]:
```suggestion
    // test for predict_words/recommendation of words based on a given prefix
```

File: data_structures/trie_using_hashmap.cpp [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+    pred_words = obj.predict_words("bu");
+
+    for (const std::string& str : pred_words) {
+        std::cout << str << std::endl;
+    }
+
+    assert(pred_words.size() == 2);
+    std::cout << "Returned all words that start with prefix bu " << std::endl;
+
+    /*tests for delete method */
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:306]:
```suggestion
    // tests for delete method
```

File: data_structures/trie_using_hashmap.cpp [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+    std::cout << "word app is deleted sucessful" << std::endl;
+
+    pred_words = obj.predict_words("app");
+    for (const std::string& str : pred_words) {
+        std::cout << str << std::endl;
+    }
+    assert(pred_words.size() == 4);
+    std::cout << "app is deleted sucessful" << std::endl;
+
+    /*test case for chinese language*/
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:319]:
```suggestion
    // test case for Chinese language
```

File: data_structures/trie_using_hashmap.cpp [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+    std::cout << "All tests passed" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run Self-test implementaions
+
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:337]:
```suggestion
    test();  // run self-test implementaions
```

File: data_structures/trie_using_hashmap.cpp [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+namespace data_structures {
+/**
+ *@brief Trie class. implementaion of trie using hashmap in each trie node for
+ *all the charters of char16_t(UTF-16)type with methods to insert, delete,
+ *search, startwith and to recommend words based on a given prefix.
+ */
+class Trie {
+ private:
+    /**
+     *  @brief struct representing nodes in trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:33]:
```suggestion
     * @brief struct representing nodes in the trie
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: data_structures/trie_using_hashmap.cpp (hunk 16 above)
Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: Thank you for helping me, I have updated the documentation as requested. I can see that the workflow has been updated recently, Will this have impact on my commit?

Comment (Panquesito7) [reply data_structures/trie_using_hashmap.cpp]: > Will this have an impact on my commit

No, everything should work perfectly fine. 🙂

File: data_structures/trie_using_hashmap.cpp [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+#include <memory>         /// for std::shared_ptr
+#include <stack>          /// for std::stack
+#include <unordered_map>  /// for std::unordered_map
+#include <vector>         /// for std::vector
+
+/**
+ * @namespace data_structures
+ * @brief Data structures algorithms
+ */
+namespace data_structures {
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:24]:
```suggestion
namespace data_structures {
/**
 * @namespace trie_using_hashmap
 * @brief Functions for [Trie](https://en.wikipedia.org/wiki/Trie) data structure using hashmap implementation
```

File: data_structures/trie_using_hashmap.cpp [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+    std::cout << "All tests passed" << std::endl;
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run self-test implementaions
+
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:335]:
```suggestion
```

File: data_structures/trie_using_hashmap.cpp [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+        }
+
+        result = get_all_words(
+            result, curr,
+            prefix);  ///< iteratively and recursively get the recommended words
+
+        return result;
+    }
+};
+
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:205]:
```suggestion
}  // namespace trie_using_hashmap
```

File: data_structures/trie_using_hashmap.cpp [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+}  // namespace data_structures
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    data_structures::Trie obj;
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:213]:
```suggestion
    data_structures::trie_using_hashmap::Trie obj;
```

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: Hi, I had Updated the code, but some checks are failing with this error in workflow
error1

File: data_structures/trie_using_hashmap.cpp [hunk 21]
```diff
@@ -0,0 +1,344 @@
...
+class Trie {
+ private:
+    /**
+     * @brief struct representing nodes in the trie
+     */
+    struct Node {
+        std::unordered_map<char16_t, std::shared_ptr<Node>>
+            children;  /// unordered map with key type char16_t and value is a
+                       /// shared pointer type of Node
+        bool word_end = false;  /// boolean variable to represent the node end
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:48]:
```suggestion
            children;  ///< unordered map with key type char16_t and value is a
                       ///< shared pointer type of Node
        bool word_end = false;  ///< boolean variable to represent the node end
```

File: data_structures/trie_using_hashmap.cpp [hunk 22]
```diff
@@ -0,0 +1,344 @@
...
+            curr = curr->children[ch];
+        }
+
+        if (!curr->word_end && curr != root_node) {
+            curr->word_end = true;
+        }
+    }
+
+    /**
+     * @brief search a word/string exists inside the trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:77]:
```suggestion
     * @brief search a word/string inside the trie
```

File: data_structures/trie_using_hashmap.cpp [hunk 23]
```diff
@@ -0,0 +1,344 @@
...
+                return false;
+            }
+            curr = curr->children[ch];
+        }
+        return true;
+    }
+
+    /**
+     * @brief delete a word/string from a trie
+     * @param word string to delete fromm trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:120]:
```suggestion
     * @param word string to delete from trie
```

File: data_structures/trie_using_hashmap.cpp [hunk 24]
```diff
@@ -0,0 +1,344 @@
...
+            }
+            if (curr->word_end) {
+                cnt++;
+            }
+
+            nodes.push(curr->children[ch]);
+            curr = curr->children[ch];
+        }
+        /*delete only when its a word and it has no children after and also no
+         * prefix in the line*/
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:138]:
```suggestion
        // Delete only when it's a word, and it has no children after
        // and also no prefix in the line
```

File: data_structures/trie_using_hashmap.cpp [hunk 25]
```diff
@@ -0,0 +1,344 @@
...
+        if (nodes.top()->word_end) {
+            nodes.top()->word_end = false;
+        }
+        while (!(nodes.top()->word_end) && nodes.top()->children.empty()) {
+            /* code */
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:143]: Please add proper documentation here.

File: data_structures/trie_using_hashmap.cpp [hunk 26]
```diff
@@ -0,0 +1,344 @@
...
+            nodes.pop();
+            nodes.top()->children.erase(word.back());
+            word.pop_back();
+        }
+    }
+
+    /**
+     * @brief helper function to predict/recommend words that starts with a
+     * given prefix from the end of prefix's node iterate through all the child
+     * nodes by recursively appending  all the possible words below the trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:153]:
```suggestion
     * nodes by recursively appending all the possible words below the trie
```

File: data_structures/trie_using_hashmap.cpp [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+     * @param prefix string to recommend the words
+     * @param element node at the end of given prefix
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:155]:
```suggestion
     * @param element node at the end of the given prefix
```

File: data_structures/trie_using_hashmap.cpp [hunk 28]
```diff
@@ -0,0 +1,344 @@
...
+                get_all_words(results, element->children[x.first], prefix);
+
+            prefix.pop_back();
+        }
+
+        return results;
+    }
+
+    /**
+     * @brief predict/recommend a words that starts with a given prefix
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:183]:
```suggestion
     * @brief predict/recommend a word that starts with a given prefix
```

File: data_structures/trie_using_hashmap.cpp [hunk 29]
```diff
@@ -0,0 +1,344 @@
...
+}  // namespace data_structures
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    data_structures::trie_using_hashmap::Trie obj;
+    /* Inserting data into trie using insert method and testing it with search
+     * method */
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:223]:
```suggestion
    // Inserting data into trie using the insert
    // method and testing it with search method
```

File: data_structures/trie_using_hashmap.cpp [hunk 30]
```diff
@@ -0,0 +1,345 @@
...
+
+    /**
+     * @brief predict/recommend a word that starts with a given prefix
+     * @param prefix string to search for
+     * @returns list of recommended words
+     */
+    std::vector<std::string> predict_words(const std::string& prefix) {
+        std::vector<std::string> result;
+        std::shared_ptr<Node> curr = root_node;
+        // treversing till the end of given prefix in trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:191]: Shouldn't here be `traversing`?
```suggestion
        // traversing until the end of the given prefix in trie
```

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp:191]: yes it's traversing, I will update it. Thank you

**Review by Panquesito7 — APPROVED** [review]
Amazing work; I like how you've made & improved the code so well! 🎉
Code and documentation look good and very refined; LGTM. 😄 👍

**Comment by ayaankhan98** [comment]
@bharath000 please mark the conversations as resolved which you have considered.

File: data_structures/trie_using_hashmap.cpp [hunk 31]
```diff
@@ -0,0 +1,345 @@
...
+
+/**
+ * @namespace trie_using_hashmap
+ * @brief Functions for [Trie](https://en.wikipedia.org/wiki/Trie) data
+ * structure using hashmap implementation
+ */
+namespace trie_using_hashmap {
+
+/**
+ * @brief Trie class. implementation of trie using hashmap in each trie node
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:34]:
```suggestion
 * @brief Trie class, implementation of trie using hashmap in each trie node
```

File: data_structures/trie_using_hashmap.cpp [hunk 32]
```diff
@@ -0,0 +1,345 @@
...
+ * for all the charters of char16_t(UTF-16)type with methods to insert,
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:35]:
```suggestion
 * for all the characters of char16_t(UTF-16)type with methods to insert,
```

File: data_structures/trie_using_hashmap.cpp [hunk 33]
```diff
@@ -0,0 +1,345 @@
...
+ * delete, search, start with, and to recommend words based on a given
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:36]:
```suggestion
 * delete, search, start with and to recommend words based on a given
```

File: data_structures/trie_using_hashmap.cpp [hunk 34]
```diff
@@ -0,0 +1,345 @@
...
+ * prefix.
+ */
+class Trie {
+ private:
+    /**
+     * @brief struct representing nodes in the trie
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:42]:
```suggestion
     * @brief struct representing a trie node.
```

File: data_structures/trie_using_hashmap.cpp [hunk 35]
```diff
@@ -0,0 +1,345 @@
...
+    std::shared_ptr<Node> root_node =
+        std::make_shared<Node>();  ///< declaring root node of trie
+
+ public:
+    ///< Constructor
+    Trie() = default;
+
+    /**
+     * @brief insert the string into the trie
+     * @param word string to insert in the tree
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:60]:
```suggestion
     * @param word string to insert in the trie
```

File: data_structures/trie_using_hashmap.cpp [hunk 36]
```diff
@@ -0,0 +1,345 @@
...
+
+    /**
+     * @brief predict/recommend a word that starts with a given prefix
+     * @param prefix string to search for
+     * @returns list of recommended words
+     */
+    std::vector<std::string> predict_words(const std::string& prefix) {
+        std::vector<std::string> result;
+        std::shared_ptr<Node> curr = root_node;
+        // traversing utill the end of the given prefix in trie
```
Comment (ayaankhan98) [inline data_structures/trie_using_hashmap.cpp:191]:
```suggestion
        // traversing till the end of the given prefix in trie
```

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: I have updated the changes, Thank you

**Comment by ayaankhan98** [comment]
@bharath000 please mark the considered changes as resolved.

File: data_structures/trie_using_hashmap.cpp [hunk 37]
```diff
@@ -0,0 +1,345 @@
...
+
+    /**
+     * @brief predict/recommend a word that starts with a given prefix
+     * @param prefix string to search for
+     * @returns list of recommended words
+     */
+    std::vector<std::string> predict_words(const std::string& prefix) {
+        std::vector<std::string> result;
+        std::shared_ptr<Node> curr = root_node;
+        // traversing till the end of the given prefix in trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:191]:
```suggestion
        // traversing until the end of the given prefix in trie
```

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: I have updated the change, ayaankhan98 requested me to change it "till" in the above request. Thank you

Comment (Panquesito7) [reply data_structures/trie_using_hashmap.cpp]: AFAIK, `utill` is wrong. It should be `until`. 🙂

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: Yes I think its until.

File: data_structures/trie_using_hashmap.cpp [hunk 38]
```diff
@@ -0,0 +1,345 @@
...
+
+    /**
+     * @brief predict/recommend a word that starts with a given prefix
+     * @param prefix string to search for
+     * @returns list of recommended words
+     */
+    std::vector<std::string> predict_words(const std::string& prefix) {
+        std::vector<std::string> result;
+        std::shared_ptr<Node> curr = root_node;
+        // traversing utill the end of the given prefix in trie
```
Comment (Panquesito7) [inline data_structures/trie_using_hashmap.cpp:191]:
```suggestion
        // traversing until the end of the given prefix in trie
```

File: data_structures/trie_using_hashmap.cpp (hunk 37 above)
Comment (ayaankhan98) [reply data_structures/trie_using_hashmap.cpp]: but i didn't suggested `untill` or `until`, i suggested `till` and i think `till` was fine.
Well i have no issues over this.

Comment (ayaankhan98) [reply data_structures/trie_using_hashmap.cpp]: everything is same either `till` or `until`

Comment (bharath000) [reply data_structures/trie_using_hashmap.cpp]: The changes have been updated. Thank you ayaankhan98 and Panquesito7, I am resolving this conversation.

**Review by ayaankhan98 — APPROVED** [review]
@bharath000 Thanks for your contribution.

**Pull request merged**